### PR TITLE
Add a default build configuration to ocb.

### PR DESCRIPTION
**Pull request by jpeach**
**Description:** 

Embed the build configuration that is used to build otelcorecol into ocb
so that end users can easily generate a useful collector. This makes the
`--config` flag optional again.

**Link to tracking Issue:**

None.

**Testing:**
```
$ make ocb
/Library/Developer/CommandLineTools/usr/bin/make -C cmd/builder config
sed '-e/replaces:/,$d' <../otelcorecol/builder-config.yaml > internal/config/default.yaml
/Library/Developer/CommandLineTools/usr/bin/make -C cmd/builder ocb
GO111MODULE=on CGO_ENABLED=0 go build -trimpath -o ../../bin/ocb_darwin_arm64 .
$ ./bin/ocb_darwin_arm64
2022-07-26T16:05:06.722+1000	INFO	internal/command.go:91	OpenTelemetry Collector Builder	{"version": "dev", "date": "unknown"}
2022-07-26T16:05:06.722+1000	INFO	internal/command.go:102	Using default build configuration
2022-07-26T16:05:06.723+1000	INFO	builder/config.go:99	Using go	{"go-executable": "/opt/homebrew/bin/go"}
2022-07-26T16:05:06.724+1000	INFO	builder/main.go:76	Sources created	{"path": "/var/folders/hf/8frlwk6s40j0wj3jpgbq94z00000gp/T/otelcol-distribution1809706932"}
2022-07-26T16:05:06.913+1000	INFO	builder/main.go:108	Getting go modules
2022-07-26T16:05:06.942+1000	INFO	builder/main.go:87	Compiling
2022-07-26T16:05:07.948+1000	INFO	builder/main.go:94	Compiled	{"binary": "/var/folders/hf/8frlwk6s40j0wj3jpgbq94z00000gp/T/otelcol-distribution1809706932/otelcorecol"}
```

**Documentation:** 

None.

### CODE REVIEW

File: cmd/builder/internal/config/default.yaml
Comment (TylerHelmuth) [inline cmd/builder/internal/config/default.yaml:6]: More of a question for Core approvers/maintainers, but I think updating this file would need to become part of the release process right?

Comment (TylerHelmuth) [reply cmd/builder/internal/config/default.yaml:6]: Would be cool if ocb was able to use the [releases manifest](https://github.com/open-telemetry/opentelemetry-collector-releases/blob/main/distributions/otelcol/manifest.yaml).

File: cmd/builder/internal/config/default.go
Comment (TylerHelmuth) [inline cmd/builder/internal/config/default.go:31]: Do we need to use FS here or could we use file.Provider instead?  I ask bc Fs is listed as experimental still.

**Review by TylerHelmuth — COMMENTED** [review]
Is it possible to add a new happy-path test to `test/test.sh` to test this new functionality?  It would be good to have a test that verifies the ocb can be run without a config.

File: cmd/builder/internal/config/default.yaml
Comment (jpeach) [reply cmd/builder/internal/config/default.yaml:6]: > updating this file would need to become part of the release process right

Yep, it gets updated as part of the `prepare-release` make target.

File: cmd/builder/internal/config/default.go
Comment (jpeach) [reply cmd/builder/internal/config/default.go:31]: Where is it marked experimental?

https://pkg.go.dev/github.com/knadh/koanf@v1.4.2/providers/fs

**Comment by jpeach** [comment]
> Is it possible to add a new happy-path test to `test/test.sh` to test this new functionality? It would be good to have a test that verifies the ocb can be run without a config.

Done

```
builder jpeach$ ./test/test.sh
Using /opt/homebrew/bin/go to compile the distributions.
Running the tests in /var/folders/hf/8frlwk6s40j0wj3jpgbq94z00000gp/T/tmp.1yTemUe3I7
✅ PASS default
✅ PASS core
```

**Comment by jpeach** [comment]
Pretty sure that this PR doesn't break contrib-tests.

File: cmd/builder/internal/config/default.yaml
Comment (TylerHelmuth) [reply cmd/builder/internal/config/default.yaml:6]: It looks like that target knows to update `/cmd/builder/internal/builder/config.go`, but would need updated to know to update these files.

File: cmd/builder/internal/config/default.go
Comment (TylerHelmuth) [reply cmd/builder/internal/config/default.go:31]: I found it marked experimental on the github page: https://github.com/knadh/koanf#bundled-providers.  That particular line hasn't been updated in 13 months, so could be out of date.

Comment (jpeach) [reply cmd/builder/internal/config/default.go:31]: I see. I was looking at the [source](https://github.com/knadh/koanf/blob/v1.4.2/providers/fs/fs.go) and godoc, which doesn't mention experimental status. We can switch to the file provider, but the code needed for that will be basically the same as the fs one, so I don't think there's much benefit.

File: cmd/builder/internal/config/default.yaml
Comment (jpeach) [reply cmd/builder/internal/config/default.yaml:6]: Sorry, I'm not following you here. `make config` in the cmd/builder directory updates cmd/builder/internal/builder/config/default.yaml, which is then embedded in cmd/builder/internal/builder/config/default.go.

Comment (TylerHelmuth) [reply cmd/builder/internal/config/default.yaml:6]: I missed that you also updated the  root Makefile 👍

**Comment by jpkrohling** [comment]
I just restarted the failed job. If it fails, please try rebasing this PR, as there might be some incompatibility problem not caused by this PR, but fixed in the main branch already (of either side).

**Comment by jpkrohling** [comment]
PR rebased.

**Comment by jpeach** [comment]
```

**Comment by jpeach** [comment]
```
# github.com/influxdata/influxdb-observability/common
Error: /home/runner/go/pkg/mod/github.com/influxdata/influxdb-observability/common@v0.2.24/metrics_sort.go:56:43: hdp.MExplicitBounds undefined (type internal.HistogramDataPoint has no field or method MExplicitBounds)
Error: /home/runner/go/pkg/mod/github.com/influxdata/influxdb-observability/common@v0.2.24/metrics_sort.go:57:21: hdp.MExplicitBounds undefined (type internal.HistogramDataPoint has no field or method MExplicitBounds)

**Comment by jpkrohling** [comment]
The failure is related to https://github.com/open-telemetry/opentelemetry-collector/issues/5770

**Review by TylerHelmuth — APPROVED** [review]

File: cmd/builder/internal/config/default.yaml
Comment (jpkrohling) [reply cmd/builder/internal/config/default.yaml:6]: Does this need to be updated now that 0.57.2 is out?

**Review by jpkrohling — COMMENTED** [review]
LGTM, but the versions need to be bumped.

**Comment by jpeach** [comment]
> LGTM, but the versions need to be bumped.

Updated

**Review by jpkrohling — APPROVED** [review]

File: CHANGELOG.md
Comment (bogdandrutu) [inline CHANGELOG.md:119]: Confused about these changes...

Comment (jpeach) [reply CHANGELOG.md:119]: Just making the formatting consistent

Comment (bogdandrutu) [reply CHANGELOG.md:119]: Please split them in a separate PR :)

Comment (jpeach) [reply CHANGELOG.md]: reverted and rebased

Comment (jpeach) [reply CHANGELOG.md]: I can rebase this again. Is it ready to merge when I do?

Comment (jpkrohling) [reply CHANGELOG.md]: @bogdandrutu, I believe this PR is pending on your side. Would you be able to review/merge this?

Comment (bogdandrutu) [reply CHANGELOG.md]: I think it should be your side as well, so we should not have "refactoring, formatting" changes in the same PR. I sent a separate PR https://github.com/open-telemetry/opentelemetry-collector/pull/6012

**Comment by bogdandrutu** [comment]
@jpkrohling can you document the motivation? Does not seem to be that useful to produce a standard collector that we already produce and release anyway?

**Comment by jpeach** [comment]
> @jpkrohling can you document the motivation? Does not seem to be that useful to produce a standard collector that we already produce and release anyway?

Originally, when you ran ocb with no configuration, it generated a collector with no plugins, which was completely useless (see my other PR which adds a way to know what plugins you have included in the collector). The original documentation for all this claimed that ocb would produce a "default collector" when run without any arguments, and this PR restores that claim, so users can build the default collector.

**Comment by jpeach** [comment]
@open-telemetry/collector-approvers is this ready to merge?

**Review by jpkrohling — APPROVED** [review]
LGTM, but I can't merge.

@bogdandrutu, could you review and merge this if you think it's ready?

**Pull request merged**
